spark3/src/main/java/org/apache/iceberg/spark/procedures/RollbackToTimestampProcedure.java

```diff
@@ -0,0 +1,102 @@
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg.spark.procedures;
+
+import org.apache.iceberg.Snapshot;
+import org.apache.iceberg.spark.procedures.SparkProcedures.ProcedureBuilder;
+import org.apache.spark.sql.catalyst.InternalRow;
+import org.apache.spark.sql.catalyst.expressions.GenericInternalRow;
+import org.apache.spark.sql.catalyst.util.DateTimeUtils;
+import org.apache.spark.sql.connector.catalog.TableCatalog;
+import org.apache.spark.sql.connector.iceberg.catalog.ProcedureParameter;
+import org.apache.spark.sql.types.DataTypes;
+import org.apache.spark.sql.types.Metadata;
+import org.apache.spark.sql.types.StructField;
+import org.apache.spark.sql.types.StructType;
+
+/**
+ * A procedure that rollbacks a table to a given point in time.
+ */
+class RollbackToTimestampProcedure extends BaseProcedure {
+
+  private static final ProcedureParameter[] PARAMETERS = new ProcedureParameter[]{
+      ProcedureParameter.required("namespace", DataTypes.StringType),
+      ProcedureParameter.required("table", DataTypes.StringType),
+      ProcedureParameter.required("timestamp", DataTypes.TimestampType)
+  };
+
+  private static final StructType OUTPUT_TYPE = new StructType(new StructField[]{
+      new StructField("previous_snapshot_id", DataTypes.LongType, false, Metadata.empty()),
+      new StructField("current_snapshot_id", DataTypes.LongType, false, Metadata.empty())
+  });
+
+  public static ProcedureBuilder builder() {
+    return new BaseProcedure.Builder<RollbackToTimestampProcedure>() {
+      @Override
+      protected RollbackToTimestampProcedure doBuild() {
+        return new RollbackToTimestampProcedure(tableCatalog());
+      }
+    };
+  }
+
+  private RollbackToTimestampProcedure(TableCatalog catalog) {
+    super(catalog);
+  }
+
+  @Override
+  public ProcedureParameter[] parameters() {
+    return PARAMETERS;
+  }
+
+  @Override
+  public StructType outputType() {
+    return OUTPUT_TYPE;
+  }
+
+  @Override
+  public InternalRow[] call(InternalRow args) {
+    String namespace = args.getString(0);
+    String tableName = args.getString(1);
+    // timestamps in Spark have nanosecond precision so this conversion is lossy
+    long timestampMillis = DateTimeUtils.toMillis(args.getLong(2));
+
+    return modifyIcebergTable(namespace, tableName, table -> {
+      Snapshot previousSnapshot = table.currentSnapshot();
+
+      table.manageSnapshots()
+          .rollbackToTime(timestampMillis)
+          .commit();
+
+      Snapshot currentSnapshot = table.currentSnapshot();
```

Comment:
Here we have the same problem as in other cases. If there is a concurrent operation after `commit`, the output may not be precise. To make it reliable, we may attach a UUID to this operation and fetch the snapshot by it. Any other ideas are welcome.

Reply:
If we are worried about concurrent operations, then we should set a validation to fail if the current snapshot ID changes. We support this in `manageSnapshots` for cherry picking, we just don't set it for rollbacks.

Reply:
It is a bit different here. Apart from validating the concurrent operations while rolling back, I am talking about what happens after the rollback is successful.  It may be that `table.currentSnapshot()` returns a snapshot that is not the same snapshot we rollbacked to. It may belong to a concurrent operation.

Reply:
I see what you mean. I don't think it is a big deal if that happens, but we can always add a callback to pass the snapshot ID that will be used back to this function.

Reply:
I have a separate issue to track this. Let's address it later.